### PR TITLE
GLOB-39083 Update cache key logic

### PR DESCRIPTION
**Pull request by GabeLambda**
One issue with the previous implementation was that if a service
changes (either altering the resource definition, cached function, etc.)
we'd still try and pull the cached resource and attempt to treat it as
if it's valid, when there's no guarantee that that's the case.

We should be able to thus use the commit SHA as a proxy for a cache
"version". This will have the side-effect of causing each new build to
invalidate its entire cache, but will be safer by default.

Note that this is probably the wrong abstraction: this idea was pulled
from microcosm-flask, but could probably be better served elsewhere (for
both usages).

### CODE REVIEW

File: microcosm_caching/tests/test_decorators.py [hunk 1]
```diff
@@ -135,11 +171,43 @@ def test_invalidates(self):
...
+                self.graph.resource_cache.get(key),
+                is_(None),
+            )
+
+        # Then validate that it was invalidated correctly
+        assert_that(self.cached_retrieve(key_id=1)["value"], is_(4))
+        assert_that(self.cached_extended_retrieve(key_id=1)["value"], is_(5))
+        assert_that(self.cached_retrieve_for(key_id=1)["values"], is_(6))
+
+    def test_invalidate_batch(self):
```
Comment (GabeLambda) [inline microcosm_caching/tests/test_decorators.py]: This test didn't exist before; adding it in here.

File: microcosm_caching/decorators.py [hunk 2]
```diff
@@ -58,18 +64,24 @@ def from_kwargs(self, kwargs) -> Dict[str, Any]:
...
+def cache_key(cache_prefix, schema, args, kwargs, version: Optional[str] = None) -> str:
     """
     Hash a key according to the schema and input args.
 
     """
     key = (schema.__name__,) + args
     key += tuple(sorted((a, b) for a, b in kwargs.items()))
 
-    return sha1(f"{cache_prefix}:{key}".encode("utf-8")).hexdigest()
+    return sha1(f"{cache_prefix}:{version}:{key}".encode("utf-8")).hexdigest()
```
Comment (GabeLambda) [inline microcosm_caching/decorators.py]: A null-specified version is fine.

File: microcosm_caching/decorators.py [hunk 3]
```diff
@@ -58,18 +64,24 @@ def from_kwargs(self, kwargs) -> Dict[str, Any]:
...
 
 
-def cached(component, schema: Type[Schema], cache_prefix: str, ttl: int = DEFAULT_TTL):
+def cached(
+    component,
+    schema: Type[Schema],
+    cache_prefix: str,
+    ttl: int = DEFAULT_TTL,
+    schema_version: Optional[str] = None,
```
Comment (GabeLambda) [inline microcosm_caching/decorators.py]: I still want the ability for someone to override the build version if needed.

File: microcosm_caching/decorators.py [hunk 4]
```diff
@@ -155,7 +170,7 @@ def cache(*args, **kwargs) -> Schema:
                 # NB: We're caching the serialized format of the resource, meaning
                 # we need to do a (wasteful) load here to enable it to be dumped correctly
                 # later on in the flow. This could probably be made more efficient
-                return schema().load(cached_resource, unknown="exclude")
+                return schema().load(cached_resource, unknown=EXCLUDE)
```
Comment (GabeLambda) [inline microcosm_caching/decorators.py]: Unrelated, but I figure it's better to use the shared constant.

File: microcosm_caching/build_info.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+"""
```
Comment (GabeLambda) [inline microcosm_caching/build_info.py]: Again, looks pretty similar to the version in microcosm-flask.

File: microcosm_caching/tests/test_build_info.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+Unit tests for build_info
+
+"""
+from hamcrest import assert_that, is_
+from microcosm.api import create_object_graph, load_from_dict
+
+
+class TestBuildInfo:
+
+    def test_cached(self):
```
Comment (afallou) [inline microcosm_caching/tests/test_build_info.py]: Could you add a test checking that caching still works if there's no build info in the config?

Comment (GabeLambda) [reply microcosm_caching/tests/test_build_info.py]: Sure.

**Pull request merged**